### PR TITLE
Qualcomm AI Engine Direct - Intermediate Tensor Dump

### PR DESCRIPTION
**Pull request by winskuo-quic**
## Summary

- Intermediate Tensor Dump enablement
- Added UT for intermediate tensor dump

### CODE REVIEW

**Comment by winskuo-quic** [comment]
Hi @cccclai,
This PR is to enable intermediate tensor dump. Please have a look.
Thanks

**Comment by winskuo-quic** [comment]
> Hi @winskuo-quic , there is an internal build failure, could you apply this patch and update again?
> 
> ```
> --- a/executorch/backends/qualcomm/runtime/targets.bzl
> +++ b/executorch/backends/qualcomm/runtime/targets.bzl
> @@ -63,5 +63,6 @@
>              "//executorch/backends/qualcomm/aot/wrappers:wrappers",
>              "//executorch/runtime/backend:interface",
>              "//executorch/runtime/core:core",
> +            "//executorch/extension/tensor:tensor",
>          ],
>      )
> ```

Hi @cccclai,
Thanks for reviewing the PR. I have just pushed a new commit that has your patch applied.

**Comment by cccclai** [comment]
Sorry I'm having issues on merging, could you rebase?

**Comment by winskuo-quic** [comment]
Hi @cccclai,
I have just rebased to mainline.
Thanks

**Review by cccclai — APPROVED** [review]

File: backends/qualcomm/runtime/QnnManager.cpp [hunk 1]
```diff
@@ -377,30 +377,27 @@ Error QnnManager::Execute(
...
-      if (fout.fail()) {
-        QNN_EXECUTORCH_LOG_ERROR(
-            "Dump tensor name: %s Failed.", QNN_VER_PTR(output_tensor)->name);
-        return Error::Internal;
-      }
-
-      fout.write(
-          static_cast<const char*>(QNN_VER_PTR(output_tensor)->clientBuf.data),
-          QNN_VER_PTR(output_tensor)->clientBuf.dataSize);
+      std::vector<exec_aten::SizesType> sizes(
```
Comment (kimishpatel) [inline backends/qualcomm/runtime/QnnManager.cpp:386]: is this dumping only the delegate outputs or the intermediate outputs as well?

Comment (winskuo-quic) [reply backends/qualcomm/runtime/QnnManager.cpp:386]: This should also dump intermediate outputs if you turn `dump_intermediate_outputs=True` when calling `generate_qnn_executorch_compiler_spec`.

Comment (kimishpatel) [reply backends/qualcomm/runtime/QnnManager.cpp:386]: I see. And right now it is logging using debug_handle=-1. ANy chance we change that?

Comment (winskuo-quic) [reply backends/qualcomm/runtime/QnnManager.cpp:386]: Sure. We will start working on it once we finished the current GA model list.

Comment (kimishpatel) [reply backends/qualcomm/runtime/QnnManager.cpp:386]: @winskuo-quic I cannot quite figure out where do you tell qnn sdk to log intermediate values in some predefined buffer. Can you point me to it?

Comment (winskuo-quic) [reply backends/qualcomm/runtime/QnnManager.cpp:386]: The key idea here is that QNN have different tensor types, such as:
`QNN_TENSOR_TYPE_NATIVE` is for any intermediate activation tensors
`QNN_TENSOR_TYPE_STATIC` for params
`QNN_TENSOR_TYPE_APP_WRITE` for inputs
`QNN_TENSOR_TYPE_APP_READ` for outputs.

Usually, we only set the final output as `QNN_TENSOR_TYPE_APP_READ`, which makes it dump the output results. To enable intermediate tensor dump, we are just setting all intermediate tensors to `QNN_TENSOR_TYPE_APP_READ`.
We did this during `node_visitor.py`: https://github.com/pytorch/executorch/blob/90ee3474f0070443d4c4c4ca0d88fb0503a98c3f/backends/qualcomm/builders/node_visitor.py#L339

Below is the PR that first introduced intermediate tensor dump:
https://github.com/pytorch/executorch/pull/1294

**Pull request closed without merging**